atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java

```diff
@@ -630,27 +630,46 @@ private void safeAppendEntry(final ApplicationEntry entry, final AppendListener
 
   private void replicate(final IndexedRaftLogEntry indexed, final AppendListener appendListener) {
     raft.checkThread();
-    appender
-        .appendEntries(indexed.index())
-        .whenCompleteAsync(
-            (commitIndex, commitError) -> {
-              if (!isRunning()) {
-                return;
-              }
+    final var appendEntriesFuture = appender.appendEntries(indexed.index());
 
-              // have the state machine apply the index which should do nothing but ensures it keeps
-              // up to date with the latest entries, so it can handle configuration and initial
-              // entries properly on fail over
-              if (commitError == null) {
-                appendListener.onCommit(indexed.index());
-                raft.notifyCommittedEntryListeners(indexed);
-              } else {
-                appendListener.onCommitError(indexed.index(), commitError);
-                // replicating the entry will be retried on the next append request
-                log.error("Failed to replicate entry: {}", indexed, commitError);
-              }
-            },
-            raft.getThreadContext());
+    if (indexed.isApplicationEntry()) {
+      // We have some services which are waiting for the application records, especially position
+      // to be committed. This is our glue code to notify them, instead of
+      // passing the complete object (IndexedRaftLogEntry) threw the listeners and
+      // keep them in heap until they are committed. This had the risk of going out of OOM
+      // if records can't be committed, see https://github.com/camunda/zeebe/issues/14275
+      final var committedPosition = indexed.getApplicationEntry().highestPosition();
+      appendEntriesFuture.whenCompleteAsync(
+          (commitIndex, commitError) -> {
+            if (!isRunning()) {
+              return;
+            }
+
+            if (commitError == null) {
+              raft.notifyApplicationEntryCommittedPositionListeners(committedPosition);
+            }
+          },
+          raft.getThreadContext());
+    }
```

Comment:
I found no better way to notify other services regarding that without passing the object. I think in general it is ok to have a specific listener for the ApplicationEntry, since we already know this interface here anyway. 

I move the future handling out, since we only want to do this when it is really about an application entry, so not in all cases.

Reply:
It's fine. It's a little confusing due to how the completable future works in terms of what gets called first - so for example, the committed position listeners may get called before the append listener, but as long as they're independent it should be fine.